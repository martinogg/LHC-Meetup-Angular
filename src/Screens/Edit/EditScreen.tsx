import React, { Component } from 'react';
import { Button, Keyboard, KeyboardAvoidingView, StyleSheet, TextInput, Text, View, Alert } from 'react-native';
import { NavigationActions, NavigationScreenProp, StackActions } from 'react-navigation'
import FirebaseConnection from '../../Helpers/FirebaseConnection'
import { IUser, User } from '../../Helpers/UserStruct'

interface IProps {
    navigation: NavigationScreenProp<any, any>,
    screenProps: {
        firebaseConnection: FirebaseConnection
    }
}

interface IState {
    userName: string,
    userLocation: string,
    userContact: string,
    userInterests: string,
}

export class EditScreen extends Component<IProps, IState> {

    constructor(props: IProps) {
        
        super(props);
        this.state = {
            userName: '',
            userLocation: '',
            userContact: '',
            userInterests: ''
        };
    }

    componentDidMount() {

        this.props.screenProps.firebaseConnection.loadUserDetails().then((user) => {

            this.setState({
                userName: user.userName,
                userLocation: user.userLocation,
                userContact: user.userContact,
                userInterests: user.userInterests
            })
        }, (error) => {

            Alert.alert('error: ' + error)
        })
    }

    saveButtonPressed() {
        
        this.props.screenProps.firebaseConnection.saveUserDetails(User.create(this.state.userName, this.state.userLocation, this.state.userContact, this.state.userInterests)).then( ()=> {

            this.props.navigation.dispatch(
                StackActions.reset({
                    index: 0,
                    actions: [NavigationActions.navigate({ routeName: 'Home' })]
                })
            )
        }, (error) => { 

            Alert.alert('Save Error: '+error)
        })

        
    }

    cancelButtonPressed() {
        // TODO cancel should go home, only if details were previously added
        this.props.navigation.dispatch(
            StackActions.reset({
                index: 0,
                actions: [NavigationActions.navigate({ routeName: 'Home' })]
            })
        )
    }

    logoutButtonPressed() {
        this.props.screenProps.firebaseConnection.logout().then(() => {

            this.props.navigation.dispatch(
                StackActions.reset({
                    index: 0,
                    actions: [NavigationActions.navigate({ routeName: 'Login' })]
                })
            )
        })
    }

    private fbTestButtonPressed() {
        // TODO - Get rid of this
        this.props.screenProps.firebaseConnection.fbTest()
    }

    public render() {
        return (
            <KeyboardAvoidingView behavior="padding" style={styles.container}>

                <View style={styles.loginContainer}>


                </View>

                <View style={styles.entriesContainer}>
                    <TextInput style={styles.input}
                        ref={(name) => this.nameInput = name}
                        autoCapitalize="none"
                        onSubmitEditing={() => this.locationInput.focus()}
                        autoCorrect={false}
                        keyboardType='default'
                        returnKeyType="next"
                        placeholder='Name'
                        value={this.state.userName}
                        placeholderTextColor='rgba(225,225,225,0.7)'
                        onChangeText={(text) => this.handleNameChange(text)}
                    />

                    <TextInput style={styles.input}
                        ref={(location) => this.locationInput = location}
                        autoCapitalize="none"
                        onSubmitEditing={() => this.contactInput.focus()}
                        autoCorrect={false}
                        keyboardType='default'
                        returnKeyType="next"
                        placeholder='Location'
                        value={this.state.userLocation}
                        placeholderTextColor='rgba(225,225,225,0.7)'
                        onChangeText={(text) => this.handleLocationChange(text)}
                    />

                    <TextInput style={styles.input}
                        ref={(contact) => this.contactInput = contact}
                        autoCapitalize="none"
                        onSubmitEditing={() => this.interestsInput.focus()}
                        autoCorrect={false}
                        keyboardType='default'
                        returnKeyType="next"
                        value={this.state.userContact}
                        placeholder='Contact: Skype or phone number'
                        placeholderTextColor='rgba(225,225,225,0.7)'
                        onChangeText={(text) => this.handleContactChange(text)}
                    />

                    <TextInput style={styles.input}
                        ref={(interests) => this.interestsInput = interests}
                        autoCapitalize="none"
                        onSubmitEditing={Keyboard.dismiss}
                        autoCorrect={false}
                        keyboardType='default'
                        returnKeyType="next"
                        placeholder='Interests'
                        value={this.state.userInterests}
                        placeholderTextColor='rgba(225,225,225,0.7)'
                        onChangeText={(text) => this.handleInterestsChange(text)}
                    />
                </View>

                <Button title="Save"
                    onPress={() => this.saveButtonPressed()}
                />
                <Button title="Cancel"
                    onPress={() => this.cancelButtonPressed()}
                />
                <Button title="Logout"
                    onPress={() => this.logoutButtonPressed()}
                />
                <Button title="Write FB Data Test"
                    onPress={() => this.fbTestButtonPressed()}
                />


            </KeyboardAvoidingView>
        );
    }

    private nameInput: any
    private locationInput: any
    private contactInput: any
    private interestsInput: any

    private handleNameChange(text: string) {

        this.setState({ userName: text })
    }

    private handleLocationChange(text: string) {

        this.setState({ userLocation: text })
    }

    private handleContactChange(text: string) {

        this.setState({ userContact: text })
    }

    private handleInterestsChange(text: string) {

        this.setState({ userInterests: text })
    }
}

const styles = StyleSheet.create({
    entriesContainer: {
        padding: 20
    },
    container: {
        flex: 1,
        backgroundColor: '#2c3e50',
    },
    loginContainer: {
        alignItems: 'center',
        flexGrow: 1,
        justifyContent: 'center'
    },
    logo: {
        position: 'absolute',
        width: 300,
        height: 100
    },
    title: {
        color: "#FFF",
        marginTop: 120,
        width: 180,
        textAlign: 'center',
        opacity: 0.9
    },
    input: {
        height: 40,
        backgroundColor: 'rgba(225,225,225,0.2)',
        marginBottom: 10,
        padding: 10,
        color: '#fff'
    }
})