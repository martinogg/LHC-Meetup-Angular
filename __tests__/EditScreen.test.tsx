import { shallow } from 'enzyme';
import React from 'react';
import 'react-native';
import { Button } from 'react-native'

import {EditScreen} from '../src/Screens/Edit/EditScreen';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

const createTestProps = (props: Object) => ({
  navigation: {
    navigate: jest.fn(),
    replace: jest.fn(),
    dispatch: jest.fn()
  },
  ...props
});

it('should display FeedbackScreen with no errors', () => {

  const props: any = createTestProps({})

  expect(renderer.create(<EditScreen {...props}/>)).toMatchSnapshot();
});

test('test save button push', () => {
        
  const navigateFunc = jest.fn()
  const navigation = { navigate: navigateFunc };

  const wrapper = shallow(<EditScreen navigation={navigation} />);
  const sut: any = wrapper.instance()

  sut.saveButtonPressed = jest.fn()

  wrapper.find(Button).at(0).simulate('press')

  expect(sut.saveButtonPressed).toHaveBeenCalledTimes(1)
});

test('test cancel push', () => {

  const navigateFunc = jest.fn()
  const navigation = { navigate: navigateFunc };

  const wrapper = shallow(<EditScreen navigation={navigation} />);
  const sut: any = wrapper.instance()

  sut.cancelButtonPressed = jest.fn()

  wrapper.find(Button).at(1).simulate('press')

  expect(sut.cancelButtonPressed).toHaveBeenCalledTimes(1)
});

test('test logout button push', () => {

  const navigateFunc = jest.fn()
  const navigation = { navigate: navigateFunc };

  const wrapper = shallow(<EditScreen navigation={navigation} />);
  const sut: any = wrapper.instance()

  sut.logoutButtonPressed = jest.fn()

  wrapper.find(Button).at(2).simulate('press')

  expect(sut.logoutButtonPressed).toHaveBeenCalledTimes(1)
});

test('test logout function', async () => {

  const logoutFunc = (): Promise<string> => {

    return new Promise((resolve, reject) => {
      resolve()
    })
  }

  let props: any;
  props = createTestProps({
    screenProps: {
      firebaseConnection: {
        logout: logoutFunc
      }
    }
  });

  const wrapper = shallow(<EditScreen {...props} />);
  const sut: any = wrapper.instance()

  await sut.logoutButtonPressed()
  expect(props.navigation.dispatch).toHaveBeenCalledTimes(1)
});