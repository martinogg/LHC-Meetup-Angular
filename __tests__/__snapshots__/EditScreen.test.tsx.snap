// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display EditScreen with no errors 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "#2c3e50",
        "flex": 1,
      },
      Object {
        "paddingBottom": 0,
      },
    ]
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexGrow": 1,
        "justifyContent": "center",
      }
    }
  />
  <View
    style={
      Object {
        "padding": 20,
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      autoCorrect={false}
      keyboardType="default"
      onChangeText={[Function]}
      onSubmitEditing={[Function]}
      placeholder="Name"
      placeholderTextColor="rgba(225,225,225,0.7)"
      rejectResponderTermination={true}
      returnKeyType="next"
      style={
        Object {
          "backgroundColor": "rgba(225,225,225,0.2)",
          "color": "#fff",
          "height": 40,
          "marginBottom": 10,
          "padding": 10,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      autoCorrect={false}
      keyboardType="default"
      onChangeText={[Function]}
      onSubmitEditing={[Function]}
      placeholder="Location"
      placeholderTextColor="rgba(225,225,225,0.7)"
      rejectResponderTermination={true}
      returnKeyType="next"
      style={
        Object {
          "backgroundColor": "rgba(225,225,225,0.2)",
          "color": "#fff",
          "height": 40,
          "marginBottom": 10,
          "padding": 10,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      autoCorrect={false}
      keyboardType="default"
      onChangeText={[Function]}
      onSubmitEditing={[Function]}
      placeholder="Contact: Skype or phone number"
      placeholderTextColor="rgba(225,225,225,0.7)"
      rejectResponderTermination={true}
      returnKeyType="next"
      style={
        Object {
          "backgroundColor": "rgba(225,225,225,0.2)",
          "color": "#fff",
          "height": 40,
          "marginBottom": 10,
          "padding": 10,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      autoCorrect={false}
      keyboardType="default"
      onChangeText={[Function]}
      onSubmitEditing={[Function]}
      placeholder="Interests"
      placeholderTextColor="rgba(225,225,225,0.7)"
      rejectResponderTermination={true}
      returnKeyType="next"
      style={
        Object {
          "backgroundColor": "rgba(225,225,225,0.2)",
          "color": "#fff",
          "height": 40,
          "marginBottom": 10,
          "padding": 10,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <View
    accessibilityRole="button"
    accessibilityStates={Array []}
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "padding": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Save
      </Text>
    </View>
  </View>
  <View
    accessibilityRole="button"
    accessibilityStates={Array []}
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "padding": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Cancel
      </Text>
    </View>
  </View>
  <View
    accessibilityRole="button"
    accessibilityStates={Array []}
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "padding": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Logout
      </Text>
    </View>
  </View>
  <View
    accessibilityRole="button"
    accessibilityStates={Array []}
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "padding": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Write FB Data Test
      </Text>
    </View>
  </View>
</View>
`;
